File: src/main/java/cc/sukazyo/cono/morny/bot/event/OnCallMsgSend.java
package cc.sukazyo.cono.morny.bot.event;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import com.pengrad.telegrambot.model.Chat;
import com.pengrad.telegrambot.model.Message;
import com.pengrad.telegrambot.model.MessageEntity;
import com.pengrad.telegrambot.model.Update;
import com.pengrad.telegrambot.model.request.ParseMode;
import com.pengrad.telegrambot.request.GetChat;
import com.pengrad.telegrambot.request.SendMessage;
import com.pengrad.telegrambot.request.SendSticker;

import cc.sukazyo.cono.morny.MornyCoeur;
import cc.sukazyo.cono.morny.bot.api.EventListener;
import cc.sukazyo.cono.morny.data.TelegramStickers;
import com.pengrad.telegrambot.response.GetChatResponse;
import com.pengrad.telegrambot.response.SendResponse;

import static cc.sukazyo.cono.morny.util.tgapi.formatting.MsgEscape.escapeHtml;


public class OnCallMsgSend extends EventListener {
	
	private static final Pattern REGEX_MSG_SENDREQ_DATA_HEAD = Pattern.compile("^\\*msg([\\d-]+)(\\*\\S+)?\\n([\\s\\S]+)$");
	
	private record MessageToSend (
			String message,
			MessageEntity[] entities,
			ParseMode parseMode,
			long targetId
	) { }
	
	@Override
	public boolean onMessage(@Nonnull Update update) {
		
		// 执行体检查
		if (update.message().chat().type() != Chat.Type.Private) return false;
		if (update.message().text() == null) return false;
		if (!update.message().text().startsWith("*msg")) return false;
		
		// 权限检查
		if (!MornyCoeur.trustedInstance().isTrusted(update.message().from().id())) {
			MornyCoeur.extra().exec(new SendSticker(
					update.message().chat().id(),
					TelegramStickers.ID_403
			).replyToMessageId(update.message().messageId()));
			return true;
		}
		
		Message msgsendReqRaw; // 用户书写的发送请求原文
		MessageToSend msgsendReqBody; // 解析后的发送请求实例
		
		// *msgsend 发送标识
		// 处理发送要求
		if (update.message().text().equals("*msgsend")) {
			// 发送体处理
			if (update.message().replyToMessage() == null) return answer404(update);
			msgsendReqBody = parseRequest(update.message().replyToMessage());
			if (msgsendReqBody == null) return answer404(update);
			// 执行发送任务
			SendResponse sendResponse = MornyCoeur.getAccount().execute(parseMessageToSend(msgsendReqBody));
			if (!sendResponse.isOk()) { // 发送失败
				MornyCoeur.extra().exec(new SendMessage(
						update.message().chat().id(),
						String.format("""
							<b><u>%d</u> FAILED</b>
							<code>%s</code>""",
								sendResponse.errorCode(),
								sendResponse.description()
						)
				).replyToMessageId(update.message().messageId()).parseMode(ParseMode.HTML));
			} else { // 发送成功信号
				MornyCoeur.extra().exec(new SendSticker(
						update.message().chat().id(),
						TelegramStickers.ID_SENT
				).replyToMessageId(update.message().messageId()));
			}
			return true;
			// 发送完成/失败 - 事件结束
		}
		
		// *msg 检查标识
		if (update.message().text().equals("*msg")) { // 处理对曾经的原文的检查
			if (update.message().replyToMessage() == null) {
				return answer404(update);
			}
			msgsendReqRaw = update.message().replyToMessage();
		} else if (update.message().text().startsWith("*msg")) { // 对接受到的原文进行检查
			msgsendReqRaw = update.message();
		} else {
			return answer404(update); // 未定义的动作
		}
		
		// 对发送请求的用户原文进行解析
		msgsendReqBody = parseRequest(msgsendReqRaw);
		if (msgsendReqBody == null) {
			return answer404(update);
		}
		
		// 输出发送目标信息
		GetChatResponse targetChatReq = MornyCoeur.getAccount().execute(new GetChat(msgsendReqBody.targetId()));
		if (!targetChatReq.isOk()) {
			MornyCoeur.extra().exec(new SendMessage(
					update.message().chat().id(),
					String.format("""
							<b><u>%d</u> FAILED</b>
							<code>%s</code>""",
							targetChatReq.errorCode(),
							targetChatReq.description()
					)
			).replyToMessageId(update.message().messageId()).parseMode(ParseMode.HTML));
		} else {
			MornyCoeur.extra().exec(new SendMessage(
					update.message().chat().id(),
					targetChatReq.chat().type() == Chat.Type.Private ? (
							String.format("""
									<i><u>%d</u>@%s</i>
									🔒 <b>%s</b> %s""",
									msgsendReqBody.targetId(),
									escapeHtml(targetChatReq.chat().type().name()),
									escapeHtml(targetChatReq.chat().firstName()+(targetChatReq.chat().lastName()==null?"":" "+targetChatReq.chat().lastName())),
									targetChatReq.chat().username()==null?
											String.format("<a href='tg://user?id=%d'>@@</a>", targetChatReq.chat().id()):
											(escapeHtml("@"+targetChatReq.chat().username()))
							)
					) : (
							String.format("""
									<i><u>%d</u>@%s</i>:::
									%s <b>%s</b>%s""",
									msgsendReqBody.targetId(),
									escapeHtml(targetChatReq.chat().type().name()),
									switch (targetChatReq.chat().type()) {
										case group -> "💭";
										case channel -> "📢";
										case supergroup -> "💬";
										default -> "⭕️";
									},
									escapeHtml(targetChatReq.chat().title()),
									targetChatReq.chat().username() != null?String.format(
											"  @%s", escapeHtml(targetChatReq.chat().username())
									):""
							)
					)
			).replyToMessageId(update.message().messageId()).parseMode(ParseMode.HTML));
		}
		// 发送文本测试
		SendResponse testSendResp = MornyCoeur.getAccount().execute(
				parseMessageToSend(msgsendReqBody, update.message().chat().id()).replyToMessageId(update.message().messageId())
		);
		if (!testSendResp.isOk()) {
			MornyCoeur.extra().exec(new SendMessage(
					update.message().chat().id(),
					String.format("""
							<b><u>%d</u> FAILED</b>
							<code>%s</code>""",
							testSendResp.errorCode(),
							testSendResp.description()
					)
			).replyToMessageId(update.message().messageId()).parseMode(ParseMode.HTML));
		}
		
		return true;
		
	}
	
	@Nullable
	private static MessageToSend parseRequest (@Nonnull Message requestBody) {
		
		final Matcher matcher = REGEX_MSG_SENDREQ_DATA_HEAD.matcher(requestBody.text());
		if (matcher.matches()) {
			long targetId = Long.parseLong(matcher.group(1));
			 ParseMode parseMode = matcher.group(2) == null ? null : switch (matcher.group(2)) {
				case "*markdown", "*md", "*m↓" -> ParseMode.MarkdownV2;
				case "*md1" -> ParseMode.Markdown;
				case "*html" -> ParseMode.HTML;
				default -> null;
			};
			final int offset = "*msg".length()+matcher.group(1).length()+(matcher.group(2)==null?0:matcher.group(2).length())+1;
			final ArrayList<MessageEntity> entities = new ArrayList<>();
			if (requestBody.entities() != null) for (MessageEntity entity : requestBody.entities()) {
				final MessageEntity parsed = new MessageEntity(entity.type(), entity.offset() - offset, entity.length());
				if (entity.url() != null) parsed.url(entity.url());
				if (entity.user() != null) parsed.user(entity.user());
				if (entity.language() != null) parsed.language(entity.language());
				entities.add(parsed);
			}
			return new MessageToSend(matcher.group(3), entities.toArray(MessageEntity[]::new), parseMode, targetId);
		}
		
		return null;
		
	}
	
	@Nonnull
	private static SendMessage parseMessageToSend (@Nonnull MessageToSend body) {
		return parseMessageToSend(body, body.targetId);
	}
	
	@Nonnull
	private static SendMessage parseMessageToSend (@Nonnull MessageToSend body, long targetId) {
		SendMessage sendingBody = new SendMessage(targetId, body.message);
		if (body.entities != null) sendingBody.entities(body.entities);
		if (body.parseMode != null) sendingBody.parseMode(body.parseMode);
		return sendingBody;
	}
	
	private static boolean answer404 (@Nonnull Update update) {
		MornyCoeur.extra().exec(new SendSticker(
				update.message().chat().id(),
				TelegramStickers.ID_404
		).replyToMessageId(update.message().messageId()));
		return true;
	}
	
}

File: src/main/scala/cc/sukazyo/cono/morny/bot/event/OnCallMsgSend.scala
package cc.sukazyo.cono.morny.bot.event

import cc.sukazyo.cono.morny.bot.api.{EventEnv, EventListener}
import cc.sukazyo.cono.morny.MornyCoeur
import cc.sukazyo.cono.morny.data.TelegramStickers
import cc.sukazyo.cono.morny.util.tgapi.TelegramExtensions.Bot.exec
import com.pengrad.telegrambot.model.{Chat, Message, MessageEntity, Update}
import com.pengrad.telegrambot.model.request.ParseMode
import com.pengrad.telegrambot.request.{GetChat, SendMessage, SendSticker}

import scala.collection.mutable.ArrayBuffer
import scala.language.postfixOps
import scala.util.matching.Regex

class OnCallMsgSend (using coeur: MornyCoeur) extends EventListener {
	
	private val REGEX_MSG_SENDREQ_DATA_HEAD: Regex = "^\\*msg(-?\\d+)(\\*\\S+)?(?:\\n([\\s\\S]+))?$"r
	
	case class MessageToSend (
		message: String|Null,
		entities: Array[MessageEntity]|Null,
		parseMode: ParseMode|Null,
		targetId: Long
	) {
		def toSendMessage (target_override: Long|Null = null): SendMessage =
			val useTarget = if target_override == null then targetId else target_override
			val sendMessage = SendMessage(useTarget, message)
			if entities ne null then sendMessage.entities(entities:_*)
			if parseMode ne null then sendMessage.parseMode(parseMode)
			sendMessage
	}
	private object MessageToSend:
		def from (raw: Message): MessageToSend = {
			raw.text match
				case REGEX_MSG_SENDREQ_DATA_HEAD(_target, _parseMode, _body) =>
					val target = _target toLong
					val parseMode: ParseMode | Null = _parseMode match
						case "*markdown" | "*md" | "*m↓" => ParseMode MarkdownV2
						case "*md1" => ParseMode Markdown
						case "*html" => ParseMode HTML
						case _ => null
					val bodyOffset = "*msg".length + _target.length + (if _parseMode eq null then 0 else _parseMode.length) + 1
					val entities = ArrayBuffer.empty[MessageEntity]
					if (raw.entities ne null) for (e <- raw.entities)
						val _parsed = MessageEntity(e.`type`, e.offset - bodyOffset, e.length)
						if e.url ne null then _parsed.url(e.url)
						if e.user ne null then _parsed.user(e.user)
						if e.language ne null then _parsed.language(e.language)
						if e.customEmojiId ne null then _parsed.language(e.language)
						entities += _parsed
					MessageToSend(_body, entities toArray, parseMode, target)
				case _ => null
		}
	
	override def onMessage (using event: EventEnv): Unit = {
		import event.update
		
		val message = update.message
		
		if message.chat.`type` != Chat.Type.Private then return;
		if message.text eq null then return;
		if !(message.text startsWith "*msg") then return;
		
		if (!(coeur.trusted isTrusted message.from.id))
			coeur.account exec SendSticker(
				message.chat.id,
				TelegramStickers ID_403
			).replyToMessageId(message.messageId)
			event.setEventOk
			return;
		
		if (message.text == "*msgsend") {
			
			if (message.replyToMessage eq null) { answer404; return }
			val messageToSend = MessageToSend from message.replyToMessage
			if ((messageToSend eq null) || (messageToSend.message eq null)) { answer404; return }
			val sendResponse = coeur.account execute messageToSend.toSendMessage()
			
			if (sendResponse isOk) {
				coeur.account exec SendSticker(
					update.message.chat.id,
					TelegramStickers ID_SENT
				).replyToMessageId(update.message.messageId)
			} else {
				coeur.account exec SendMessage(
					update.message.chat.id,
					// language=html
					s"""<b><u>${sendResponse.errorCode} FAILED</u></b>
					   |<code>${sendResponse.description}</code>"""
					.stripMargin
				).replyToMessageId(update.message.messageId).parseMode(ParseMode HTML)
			}
			
			event.setEventOk
			return
			
		}
		
		val messageToSend: MessageToSend =
			val raw: Message =
				if (message.text == "*msg")
					if message.replyToMessage eq null then { answer404; return }
					else message.replyToMessage
				else if (message.text startsWith "*msg")
					message
				else { answer404; return }
			val _toSend = MessageToSend from raw
			if _toSend eq null then { answer404; return }
			else _toSend
		
		val targetChatResponse = coeur.account execute GetChat(messageToSend.targetId)
		if (targetChatResponse isOk) {
			def getChatDescriptionHTML (chat: Chat): String =
				import cc.sukazyo.cono.morny.util.tgapi.formatting.TelegramFormatter.*
				import cc.sukazyo.cono.morny.util.tgapi.formatting.TelegramParseEscape.escapeHtml as h
				// language=html
				s"""<i><u>${h(chat.id toString)}</u>@${h(chat.`type`.name)}</i>${if (chat.`type` != Chat.Type.Private) ":::" else ""}
				   |${chat.typeTag} <b>${h(chat.safe_name)}</b> ${chat.safe_linkHTML}"""
				.stripMargin
			coeur.account exec SendMessage(
				update.message.chat.id,
				getChatDescriptionHTML(targetChatResponse.chat)
			).parseMode(ParseMode HTML).replyToMessageId(update.message.messageId)
		} else {
			coeur.account exec SendMessage(
				update.message.chat.id,
				// language=html
				s"""<b><u>${targetChatResponse.errorCode} FAILED</u></b>
				   |<code>${targetChatResponse.description}</code>"""
						.stripMargin
			).parseMode(ParseMode HTML).replyToMessageId(update.message.messageId)
		}
		
		if messageToSend.message eq null then { answer404; return }
		val testSendResponse = coeur.account execute
			messageToSend.toSendMessage(update.message.chat.id).replyToMessageId(update.message.messageId)
		if (!(testSendResponse isOk))
			coeur.account exec SendMessage(
				update.message.chat.id,
				// language=html
				s"""<b><u>${testSendResponse.errorCode}</u> FAILED</b>
				   |<code>${testSendResponse.description}</code>"""
				.stripMargin
			).parseMode(ParseMode HTML).replyToMessageId(update.message.messageId)
		
		event.setEventOk
		
	}
	
	private def answer404 (using event: EventEnv): Unit =
		coeur.account exec SendSticker(
			event.update.message.chat.id,
			TelegramStickers ID_404
		).replyToMessageId(event.update.message.messageId)
		event.setEventOk
	
}
